### ISSUE TITLE
AppendSuffix does not work as expected

### ISSUE THREAD

**Issue by SimonKoegler** [elastic/elasticsearch-net#2331]
**NEST/Elasticsearch.Net version**: 2.4.6

**Description of the problem including expected versus actual behavior**:
When chaining multiple `AppendSuffix(string suffix)` calls, the suffixes are appended in reverse order.
For example `expression.AppendSuffix("raw").AppendSuffix("evendeeper")` will append "evendeeper.raw" instead of "raw.evendeeper".

There is also a closely related problem; if trying to append a suffix with `AppendSuffix(string suffix)` to expressions like `p => p.Metadata["hardcoded"]` an exception is thrown:

> System.ArgumentException was caught
>   _message=Die auf Typ "System.Collections.Generic.IDictionary'2[System.String,System.String]" deklarierte Methode "System.String get_Item(System.String)" kann nicht mit einer Instanz vom Typ "System.Object" aufgerufen werden.
> 
>    bei System.Linq.Expressions.Expression.ValidateCallInstanceType(Type instanceType, MethodInfo method)
>    bei System.Linq.Expressions.Expression.ValidateStaticOrInstanceMethod(Expression instance, MethodInfo method)
>    bei System.Linq.Expressions.Expression.Call(Expression instance, MethodInfo method, IEnumerable'1 arguments)
>    bei System.Linq.Expressions.InstanceMethodCallExpressionN.Rewrite(Expression instance, IList'1 args)
>    bei System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
>    bei System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
>    bei System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
>    bei Nest.ExpressionExtensions.AppendSuffix[T](Expression'1 expression, String suffix)

**Possible solution**:
After some research I might have found the problem:
In [ExpressionExtensions](https://github.com/elastic/elasticsearch-net/blob/master/src/Nest/CommonAbstractions/Extensions/ExpressionExtensions.cs), the `SuffixExpressionVisitor` should override `Visit` instead of `VisitMember`. So the corrected class should look like this:

```
    /// <summary>
    /// Calls <see cref="SuffixExtensions.Suffix"/> on a member expression.
    /// </summary>
    private class SuffixExpressionVisitor : ExpressionVisitor
    {
        private readonly string suffix;

        public SuffixExpressionVisitor(string suffix)
        {
            this.suffix = suffix;
        }

        public override Expression Visit(Expression node)
        {
            return Expression.Call(
                typeof(SuffixExtensions),
                nameof(SuffixExtensions.Suffix),
                null,
                node,
                Expression.Constant(suffix));
        }

        protected override Expression VisitUnary(UnaryExpression node)
        {
            return node;
        }
    }
```

You might find these two tests useful, too:

```
        public void AppendingSuffixToExpressions() 
        {
            /** Here we have a list of expressions */
            var expressions = new List<Expression<Func<Project, object>>>
            {
                p => p.Name,
                p => p.Description,
                p => p.CuratedTags.First().Name,
                p => p.LeadDeveloper.FirstName,
                p => p.Metadata["hardcoded"]
            };

            /** and we want to append the suffix "raw" to each */
            var fieldExpressions =
                expressions.Select<Expression<Func<Project, object>>, Field>(e => e.AppendSuffix("raw")).ToList();


            Expect("name.raw").WhenSerializing(fieldExpressions[0]);
            Expect("description.raw").WhenSerializing(fieldExpressions[1]);
            Expect("curatedTags.name.raw").WhenSerializing(fieldExpressions[2]);
            Expect("leadDeveloper.firstName.raw").WhenSerializing(fieldExpressions[3]);
            Expect("metadata.hardcoded.raw").WhenSerializing(fieldExpressions[4]);
        }

        public void AppendingSuffixInDepthToExpressions() 
        {
            /** Here we have a list of expressions */
            var expressions = new List<Expression<Func<Project, object>>>
            {
                p => p.Name,
                p => p.Description,
                p => p.CuratedTags.First().Name,
                p => p.LeadDeveloper.FirstName
            };

            /** and we want to append the suffixes "raw" and "evendeeper" to each */
            var fieldExpressions =
                expressions.Select<Expression<Func<Project, object>>, Field>(e => e.AppendSuffix("raw").AppendSuffix("evendeeper")).ToList();


            Expect("name.raw.evendeeper").WhenSerializing(fieldExpressions[0]);
            Expect("description.raw.evendeeper").WhenSerializing(fieldExpressions[1]);
            Expect("curatedTags.name.raw.evendeeper").WhenSerializing(fieldExpressions[2]);
            Expect("leadDeveloper.firstName.raw.evendeeper").WhenSerializing(fieldExpressions[3]);
        }
```

Hope this helps!

**Comment by russcam** [comment]
@SimonKoegler thanks for reporting this :+1: 

Could you open a PR against the 2.x branch with the fix and tests?

**Comment by SimonKoegler** [comment]
I would like to, but unfortunately I don't know how to do this. Could you explain me the steps?

**Comment by russcam** [comment]
@SimonKoegler [the github documentation for PRs is pretty comprehensive](https://help.github.com/categories/collaborating-with-issues-and-pull-requests/) but in a nutshell

1. [Fork the elastic/elasticsearch-net repository](https://help.github.com/articles/fork-a-repo/) to your account
2. Checkout 2.x branch
3. Make changes to the branch and commit. If you have more than one commit to make the change, please could I ask that the [commits are squashed into one commit](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html) in order to keep the git history a little cleaner. Additionally, please add a meaningful git message such as

    > Fix AppendSuffix to work with different expressions
    >
    > AppendSuffix should work with property indexer expressions in addition to member access expressions
    > Closes #2331

    the blank space between the initial title and the rest of the commit message are important

4. Open a pull request from your 2.x branch with the change to the elastic/elasticsearch-net branch

In order to commit to the repository, you will also need to [sign the CLA agreement](https://www.elastic.co/contributor-agreement), which you will be prompted to do when you open the PR; this is nothing nefarious, it simply enables Elastic to distribute your code contribution without restriction, an agreement that's pretty important for Open Source Software 😄 

We can then review the PR, pull it in and merge it to other upstream branches.
